sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala

```diff
@@ -176,44 +176,33 @@ private[sql] object JsonInferSchema {
   }
 
   /**
-   * Canonicalize inferred types, e.g., convert NullType to StringType and remove StructTypes
-   * with no fields.
+   * Recursively canonicalizes inferred types, e.g., removes StructTypes with no fields,
+   * drops NullTypes or converts them to StringType based on provided options.
    */
   private def canonicalizeType(tpe: DataType, options: JSONOptions): Option[DataType] = tpe match {
-    case at @ ArrayType(elementType, _) =>
-      val canonicalizeArrayOption = for {
-        canonicalType <- canonicalizeType(elementType, options)
-      } yield {
-        at.copy(canonicalType)
-      }
-
-      canonicalizeArrayOption.map { array =>
-        if (options.dropFieldIfAllNull && array.elementType == NullType) {
-          NullType
-        } else {
-          array
-        }
-      }
+    case at: ArrayType =>
+      canonicalizeType(at.elementType, options)
+        .map(t => at.copy(elementType = t))
```

Comment:
We still need this copy?

Reply:
Yes, this might change ArrayType(ArrayType(NullType)) to ArrayType(ArrayType(StringType))